src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -573,25 +627,30 @@ static bool _rasterImage(SwSurface* surface, uint32_t *img, uint32_t w, TVG_UNUS
     return true;
 }
 
-
 static bool _rasterImage(SwSurface* surface, const uint32_t *img, uint32_t w, uint32_t h, const SwBBox& region, const Matrix* invTransform)
 {
     for (auto y = region.min.y; y < region.max.y; ++y) {
         auto dst = &surface->buffer[y * surface->stride + region.min.x];
         auto ey1 = y * invTransform->e12 + invTransform->e13;
         auto ey2 = y * invTransform->e22 + invTransform->e23;
         for (auto x = region.min.x; x < region.max.x; ++x, ++dst) {
-            auto rX = static_cast<uint32_t>(roundf(x * invTransform->e11 + ey1));
-            auto rY = static_cast<uint32_t>(roundf(x * invTransform->e21 + ey2));
+            auto fX = x * invTransform->e11 + ey1;
+            auto fY = x * invTransform->e21 + ey2;
+            auto rX = static_cast<uint32_t>(roundf(fX));
+            auto rY = static_cast<uint32_t>(roundf(fY));
             if (rX >= w || rY >= h) continue;
-            auto src = img[rX + (rY * w)];    //TODO: need to use image's stride
+            uint32_t src;
+            if (rX == w - 1 || rY == h - 1 || invTransform->e11 < 0 || invTransform->e22 < 0) {
```

Comment:
I understand this part ```rX == w - 1 || rY == h - 1```, since it's obvious, but this one:
```invTransform->e11 < 0 || invTransform->e22 < 0``` ?
earlier you've made a comment, that it has something to do with downscaling, but it doesn't. not in this form. 
the result below is for the pict transformed as follows: 
```
        picture->translate(600, 0);
        picture->rotate(91);   // <- cos(91) < 0
```
![pyt](https://user-images.githubusercontent.com/67589014/126478363-11982d00-f318-4b12-b9fd-d82989d986ce.PNG)

on the right I removed the condition ```invTransform->e11 < 0 || invTransform->e22 < 0```

Reply:
for sx = sy we could use ```e11^2 + e21^2 < 1``` condition to check whether it's downscaling